charms/worker/k8s/src/charm.py

```diff
@@ -234,12 +237,49 @@ def _create_cos_tokens(self):
     @on_error(
         WaitingStatus("Waiting for enable components"), InvalidResponseError, K8sdConnectionError
     )
-    def _enable_components(self):
+    def _configure_components(self):
         """Enable necessary components for the Kubernetes cluster."""
-        status.add(ops.MaintenanceStatus("Enabling DNS"))
-        self.api_manager.enable_component("dns", True)
+        if self._dns_charm_integrated():
+            status.add(ops.MaintenanceStatus("Disabling DNS"))
+            self.api_manager.configure_component("dns", False)
+            self._configure_dns()
+        else:
+            status.add(ops.MaintenanceStatus("Enabling DNS"))
+            self.api_manager.configure_component("dns", True)
+
         status.add(ops.MaintenanceStatus("Enabling Network"))
-        self.api_manager.enable_component("network", True)
+        self.api_manager.configure_component("network", True)
+
+    def _dns_charm_integrated(self) -> bool:
+        """Check if the DNS charm is integrated.
+
+        Returns:
+            bool: True if the DNS charm is integrated, False otherwise.
+        """
+        dns_relation = self.model.get_relation("dns-provider")
+        if not dns_relation:
+            return False
+        return True
+
+    def _configure_dns(self):
+        """Configure DNS with dns config from the dns-provider relation."""
+        if isinstance(self.unit.status, BlockedStatus):
+            return
+
+        dns_relation = self.model.get_relation("dns-provider")
+        if not dns_relation:
+            return
+
+        dns_ip = self.kube_dns.address
+        dns_domain = self.kube_dns.domain
+        # port = self.kube_dns.port or 53
+
+        if not dns_ip or not dns_domain:
+            return
+
+        self.unit.status = MaintenanceStatus("configuring DNS")
+        self.api_manager.configure_dns(dns_domain, dns_ip)
```

Comment:
Apply to the API manager in one spot
```suggestion
        if settings:
            self.api_manager.configure_component("dns", False)
            self.api_manager.configure_dns(settings.dns_domain, settings.dns_ip)
        else:
            self.api_manager.configure_component("dns", True)
```